### PR TITLE
Raise an exception when self test returns non zero code

### LINKED ISSUE #668: self_test() should raise an exception in the event of a self-test failure
Session.self_test() returns an int and a tuple. This is directly derived from how the C API works.

During API reviews it was pointed out that for a Python program, it's very easy to miss the return values by calling:

    session.self_test()

rather than

    code, msg = session.self_test()

It was suggested that self_test() should not return anything, instead it should raise a new exception type (SelfTestFailure or something like that).

### PR DESCRIPTION
**Pull request by texasaggie97**
- [X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
- [X] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.
- [X] I've added & updated tests applicable for this pull request
### What does this Pull Request accomplish?
* Add wrapper function around generated self test
* Raise new self test exception if non zero code returned
* Fixes tests
* Add self_test to nifake to allow unit testing
### List issues fixed by this Pull Request below, if any.

* Fix #668 

### What testing has been done?
* Unit
* System

### CODE REVIEW

File: src/nifake/unit_tests/test_session.py [hunk 1]
```diff
@@ -157,6 +157,28 @@ def test_simple_function(self):
...
+        self.patched_library.niFake_self_test.side_effect = self.side_effects_helper.niFake_self_test
+        test_error_code = 1
+        test_error_message = 'error message'
+        self.side_effects_helper['self_test']['selfTestResult'] = test_error_code
+        self.side_effects_helper['self_test']['selfTestMessage'] = test_error_message
+        with nifake.Session('dev1') as session:
+            try:
+                session.self_test()
+                assert False
+            except Exception as e:
```
Comment (texasaggie97) [inline src/nifake/unit_tests/test_session.py]: I am not sure what to do here. We do not export the individual exceptions from errors, and the other exceptions inherit from `Exception`, not the driver `Error`. I see two ways to fix this:
* Export all the different exceptions
* All driver specific exceptions inhering from the base driver exception called `Error`

What do others think?

Comment (texasaggie97) [reply src/nifake/unit_tests/test_session.py]:
``` python
import nifake.errors

try:
    blah()
except nifake.errors.SelfTestFailureError as e:
    print(e)
```
works.

Is this acceptable for customers?

File: build/templates/errors.py.mako [hunk 2]
```diff
@@ -73,6 +73,15 @@ class InvalidRepeatedCapabilityError(Exception):
         super(InvalidRepeatedCapabilityError, self).__init__('An invalid character ({0}) was found in repeated capability string ({1})'.format(invalid_character, invalid_string))
 
 
+class SelfTestFailureError(Exception):
+    '''An error due to a failed self test'''
```
Comment (marcoskirsch) [inline build/templates/errors.py.mako]: I believe tech comm says to use "self-test".

File: build/templates/session.py/self_test.py.mako [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<%page args="f, config, method_template"/>\
+<%
+    '''Call self test and throw if there is a failure'''
```
Comment (marcoskirsch) [inline build/templates/session.py/self_test.py.mako]: self-test

File: build/templates/errors.py.mako [hunk 4]
```diff
@@ -73,6 +73,15 @@ class InvalidRepeatedCapabilityError(Exception):
         super(InvalidRepeatedCapabilityError, self).__init__('An invalid character ({0}) was found in repeated capability string ({1})'.format(invalid_character, invalid_string))
 
 
+class SelfTestFailureError(Exception):
```
Comment (marcoskirsch) [inline build/templates/errors.py.mako]: I think having both "failure" and "error" (double negative) in the name of the class is redundant. One should be enough. The irony that the second sentence I typed is redundant is not lost on me.

File: generated/nifake/unit_tests/test_session.py [hunk 5]
```diff
@@ -158,6 +159,28 @@ def test_simple_function(self):
...
+        test_error_message = 'error message'
+        self.side_effects_helper['self_test']['selfTestResult'] = test_error_code
+        self.side_effects_helper['self_test']['selfTestMessage'] = test_error_message
+        with nifake.Session('dev1') as session:
+            try:
+                session.self_test()
+                assert False
+            except nifake.errors.SelfTestFailureError as e:
+                assert e.code == test_error_code
+                assert e.message == test_error_message
```
Comment (marcoskirsch) [inline generated/nifake/unit_tests/test_session.py]: You need to fail the test when an exception other than nifake.errors.SelfTestFailureError is thrown as well. As written we would get a false positive.

File: src/nidcpower/metadata/functions_addon.py [hunk 6]
```diff
@@ -148,8 +155,46 @@
             'description': 'Returns the number of channels.',
         },
     },
+    # Public function that wraps self_test and will raise on self test failure
+    'self_test_wrapper': {
```
Comment (marcoskirsch) [inline src/nidcpower/metadata/functions_addon.py]: I truly feel this should be Fancy Self Test.

Comment (texasaggie97) [reply src/nidcpower/metadata/functions_addon.py]: Changed

**Pull request merged**